src/dstack/_internal/server/background/tasks/process_runs.py

```diff
@@ -0,0 +1,283 @@
+from dstack._internal.core.models.profiles import ProfileRetryPolicy
+from dstack._internal.core.models.runs import (
+    JobStatus,
+    JobTerminationReason,
+    RunSpec,
+    RunStatus,
+    RunTerminationReason,
+)
+from dstack._internal.server.db import get_session_ctx
+from dstack._internal.server.models import InstanceModel, JobModel, RunModel
+from dstack._internal.server.services.jobs import (
+    RUNNING_PROCESSING_JOBS_IDS,
+    RUNNING_PROCESSING_JOBS_LOCK,
+    SUBMITTED_PROCESSING_JOBS_IDS,
+    SUBMITTED_PROCESSING_JOBS_LOCK,
+    TERMINATING_PROCESSING_JOBS_IDS,
+    TERMINATING_PROCESSING_JOBS_LOCK,
+)
+from dstack._internal.server.services.runs import (
+    PROCESSING_RUNS_IDS,
+    PROCESSING_RUNS_LOCK,
+    create_job_model_for_new_submission,
+    fmt,
+    process_terminating_run,
+    run_model_to_run,
+)
+from dstack._internal.server.utils.common import wait_unlock
+from dstack._internal.utils.common import get_current_datetime
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+PROCESSING_INTERVAL = datetime.timedelta(seconds=2)
+JOB_ERROR_CODES_TO_RETRY = {
+    JobTerminationReason.INTERRUPTED_BY_NO_CAPACITY,
+    JobTerminationReason.FAILED_TO_START_DUE_TO_NO_CAPACITY,
+}
+
+
+async def process_runs():
+    async with get_session_ctx() as session:
+        async with PROCESSING_RUNS_LOCK:
+            res = await session.execute(
+                sa.select(RunModel).where(
+                    RunModel.status.not_in(RunStatus.finished_statuses()),
+                    RunModel.last_processed_at < get_current_datetime() - PROCESSING_INTERVAL,
+                    RunModel.id.not_in(PROCESSING_RUNS_IDS),
+                )
+            )
+            runs: List[RunModel] = res.scalars().all()
+            PROCESSING_RUNS_IDS.update(run.id for run in runs)
+
+    futures = [process_single_run(run.id, [job.id for job in run.jobs]) for run in runs]
+    for future in asyncio.as_completed(futures):
+        try:
+            run_id = await future
+            PROCESSING_RUNS_IDS.remove(run_id)  # unlock job processing as soon as possible
+        except Exception as e:
+            logger.error("Unexpected run processing error", exc_info=e)
+
+    PROCESSING_RUNS_IDS.difference_update(
+        run.id for run in runs
+    )  # ensure that all runs are unlocked
+
+
+async def process_single_run(run_id: uuid.UUID, job_ids: List[uuid.UUID]) -> uuid.UUID:
+    jobs_ids_set = set(job_ids)
+    await wait_unlock(SUBMITTED_PROCESSING_JOBS_LOCK, SUBMITTED_PROCESSING_JOBS_IDS, jobs_ids_set)
+    await wait_unlock(RUNNING_PROCESSING_JOBS_LOCK, RUNNING_PROCESSING_JOBS_IDS, jobs_ids_set)
+    await wait_unlock(
+        TERMINATING_PROCESSING_JOBS_LOCK, TERMINATING_PROCESSING_JOBS_IDS, jobs_ids_set
+    )
+
+    async with get_session_ctx() as session:
+        run = await session.get(RunModel, run_id)
+        if run is None:
+            raise ValueError(f"Run {run_id} not found")
```

Comment:
Do not raise general-purpose ValueError. Handle the error case appropriately, e.g. ignore/log/mark run as failed. If needed to propagate to the caller, raise a custom error.